cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/LeaderReplicaDistributionGoal.java

```diff
@@ -0,0 +1,336 @@
+        Replica destinationReplica = destinationBroker.replica(action.destinationTopicPartition());
+        Broker sourceBroker = sourceReplica.broker();
+        if (sourceReplica.isLeader() && !destinationReplica.isLeader()) {
+          return (isLeaderReplicaCountUnderBalanceUpperLimitAfterChange(destinationBroker, ADD)
+                 && isLeaderReplicaCountAboveBalanceLowerLimitAfterChange(sourceBroker, REMOVE)) ? ACCEPT : REPLICA_REJECT;
+        } else if (!sourceReplica.isLeader() && destinationReplica.isLeader()) {
+          return (isLeaderReplicaCountUnderBalanceUpperLimitAfterChange(sourceBroker, ADD)
+                 && isLeaderReplicaCountAboveBalanceLowerLimitAfterChange(destinationBroker, REMOVE)) ? ACCEPT : REPLICA_REJECT;
+        }
+        return ACCEPT;
+      case INTER_BROKER_REPLICA_MOVEMENT:
+        if (sourceReplica.isLeader()) {
+          return (isLeaderReplicaCountUnderBalanceUpperLimitAfterChange(destinationBroker, ADD)
+                  && isLeaderReplicaCountAboveBalanceLowerLimitAfterChange(sourceReplica.broker(), REMOVE)) ? ACCEPT : REPLICA_REJECT;
+        }
+        return ACCEPT;
+      case LEADERSHIP_MOVEMENT:
+        return (isLeaderReplicaCountUnderBalanceUpperLimitAfterChange(destinationBroker, ADD)
+               && isLeaderReplicaCountAboveBalanceLowerLimitAfterChange(sourceReplica.broker(), REMOVE)) ? ACCEPT : REPLICA_REJECT;
+      default:
+        throw new IllegalArgumentException("Unsupported balancing action " + action.balancingAction() + " is provided.");
+    }
+  }
+
+  private boolean isLeaderReplicaCountUnderBalanceUpperLimitAfterChange(Broker broker, ReplicaDistributionGoal.ChangeType changeType) {
+    int numLeaderReplicas = broker.leaderReplicas().size();
+    return changeType == ADD ? numLeaderReplicas + 1 <= _balanceUpperLimit : numLeaderReplicas - 1 <= _balanceUpperLimit;
+  }
+
+  private boolean isLeaderReplicaCountAboveBalanceLowerLimitAfterChange(Broker broker, ReplicaDistributionGoal.ChangeType changeType) {
+    int numLeaderReplicas = broker.replicas().size();
+    return changeType == ADD ? numLeaderReplicas + 1 >= _balanceLowerLimit : numLeaderReplicas - 1 >= _balanceLowerLimit;
+  }
+
+  @Override
+  public ClusterModelStatsComparator clusterModelStatsComparator() {
+    return new LeaderReplicaDistributionGoalStatsComparator();
+  }
+
+  @Override
+  public ModelCompletenessRequirements clusterModelCompletenessRequirements() {
+    return new ModelCompletenessRequirements(MIN_NUM_VALID_WINDOWS_FOR_SELF_HEALING, 0.0, true);
+  }
+
+  @Override
+  public String name() {
+    return LeaderReplicaDistributionGoal.class.getSimpleName();
+  }
+
+  @Override
+  public boolean isHardGoal() {
+    return false;
+  }
+
+  /**
+   * Get brokers that the rebalance process will go over to apply balancing actions to replicas they contain.
+   * Leadership movement is unable to move replica off dead broker, therefore this goal only go over alive brokers.
```

Comment:
Is it possible to make this goal consistent with `ReplicaDistributionGoal` and `TopicReplicaDistributionGoal` -- e.g. can we add `_selfHealingDeadBrokersOnly` to move replicas off of dead brokers. It seems this goal fails to fix offline replicas as it is implemented.

Reply:
I added this back. But instead of `move replicas off of dead brokers` this goal can only `move **leader** replica off dead brokers` since this goal is all about leader replica distribution, it does not have enough information to make decision about follower replica movement.

Reply:
Is it possible for the leader of any partition to reside on a dead broker/disk when `RF>1`? Can't we move any replica off of the failed brokers/disks if `_selfHealingDeadBrokersOnly`?